app/models/upload.rb

```diff
@@ -70,28 +76,76 @@ def count_contents(upload, meta)
     else
       upload_path = submission_path.to_s
     end
+    output, err, status = "", "", ""
     begin
       if meta[:mimetype] == "application/zip" || upload_path.ends_with?(".zip")
-        return ZipRuby::Archive.open(upload_path).num_files
+        offset = 5 # archive name, column headers, dashes, dashes, and summary lines
+        output, status = Open3.capture2("unzip", "-l", upload_path)
       elsif meta[:mimetype] == "application/x-tar" || upload_path.ends_with?(".tar")
-        return SubTarball.count_files(upload_path)
+        output, err, status = Open3.capture3("tar", "-tvf", upload_path)
+        offset = 0
       elsif meta[:mimetype] == "application/x-compressed-tar" || upload_path.ends_with?(".tgz")
-        return SubTarball.count_files_gz(upload_path)
+        output, err, status = Open3.capture3("tar", "-tzvf", upload_path)
+        offset = 0
       elsif meta[:mimetype] == "application/gzip" || upload_path.ends_with?(".gz")
         if submission_path.to_s.ends_with?(".tar.gz")
-          return SubTarball.count_files_gz(upload_path)
+          output, err, status = Open3.capture3("tar", "-tf", upload_path)
+          offset = 0
         else
           return 1
         end
       else
         return 1
       end
+      raise Exception.new("Could not list the files in #{upload_path}") if !status.success?
+      return output.lines.count - offset
     rescue Exception => e
       puts e.message
       raise Exception.new("Could not read archive to count files")
     end
   end
 
+  def total_size_too_large?(upload, meta)
```

Comment:
An alternate plan to consider is making a 20MB RAM disk and unzipping there. If it fails, it was too big. This has the advantages of being archive format independent and avoids the two step "check and then unzip" process which is prone to logic errors.